reference/test/preconditioner/schwarz.cpp

```diff
@@ -0,0 +1,225 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/preconditioner/schwarz.hpp>
+
+
+#include <algorithm>
+#include <type_traits>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+#include <ginkgo/core/preconditioner/jacobi.hpp>
+#include <ginkgo/core/solver/cg.hpp>
+#include <ginkgo/core/solver/ir.hpp>
+#include <ginkgo/core/stop/residual_norm.hpp>
+
+
+#include "core/base/extended_float.hpp"
+#include "core/test/utils.hpp"
+
+
+namespace {
+
+
+template <typename ValueIndexType>
+class Schwarz : public ::testing::Test {
+protected:
+    using value_type =
+        typename std::tuple_element<0, decltype(ValueIndexType())>::type;
+    using index_type =
+        typename std::tuple_element<1, decltype(ValueIndexType())>::type;
+    using CsrMat = gko::matrix::Csr<value_type, index_type>;
+    using Prec = gko::preconditioner::Schwarz<value_type, index_type>;
+    using InnerSolver = gko::preconditioner::Jacobi<value_type, index_type>;
+    using Mtx = gko::matrix::Csr<value_type, index_type>;
+    using Vec = gko::matrix::Dense<value_type>;
+    using mdata = gko::matrix_data<value_type, index_type>;
+
+    Schwarz()
+        : exec(gko::ReferenceExecutor::create()),
+          mtx(gko::matrix::Csr<value_type, index_type>::create(
+              exec, gko::dim<2>{5}, 13))
+    {
+        inner_factory =
+            gko::share(InnerSolver::build().with_max_block_size(5u).on(exec));
+        schwarz_factory = gko::share(
+            Prec::build()
+                .with_subdomain_sizes(std::vector<gko::size_type>{2, 2, 1})
```

Comment:
It might make sense to try out with diverging `subdomain_sizes` and `num_subdomains` values. As well as try some edge cases (e.g., only one subdomain).